File: strategy_core/runtime.py
"""Runtime metadata and engine-clock interfaces shared by strategy contexts."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import TYPE_CHECKING, Literal, Protocol, runtime_checkable

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable
    from datetime import date, datetime

MarketType = Literal["high", "low"]


class RuntimeMode(StrEnum):
    """Known runtime modes for the shared contract."""

    PAPER = "paper"
    REPLAY = "replay"
    LIVE = "live"


@dataclass(frozen=True, slots=True)
class StrategyScope:
    """Scope facts for one running strategy sleeve."""

    sleeve_id: str
    strategy_name: str
    station_id: str | None = None
    tickers: tuple[str, ...] = field(default_factory=tuple)
    market_type: MarketType | None = None
    event_ticker: str | None = None
    event_date: date | None = None


@runtime_checkable
class TimerHandle(Protocol):
    """Handle for a scheduled wake that may be cancelled by the strategy/runtime."""

    @property
    def cancelled(self) -> bool: ...

    def cancel(self) -> None: ...


@runtime_checkable
class WorkHandle(Protocol):
    """Handle for bounded immediate work owned by the runtime."""

    @property
    def cancelled(self) -> bool: ...

    @property
    def done(self) -> bool: ...

    @property
    def exception(self) -> BaseException | None: ...

    def cancel(self) -> None: ...


@runtime_checkable
class EngineClock(Protocol):
    """Engine-owned clock so strategy logic can stay portable across runtimes."""

    def now(self) -> datetime: ...

    async def sleep(self, seconds: float) -> None: ...

    async def sleep_until(self, when: datetime) -> None: ...


@runtime_checkable
class StrategyRuntime(Protocol):
    """Runtime metadata, timer, and bounded work surface exposed to strategies."""

    mode: RuntimeMode
    run_id: str
    scope: StrategyScope
    clock: EngineClock

    def wake_at(self, when: datetime, *, name: str | None = None) -> TimerHandle: ...

    def start_work(
        self,
        work: Callable[[], Awaitable[None]],
        *,
        name: str | None = None,
    ) -> WorkHandle: ...

File: native/strategy_core/src/runtime.rs
use std::future::Future;

use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};

use crate::models::JsonValue;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum MarketType {
    High,
    Low,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RuntimeMode {
    Paper,
    Replay,
    Live,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct StrategyScope {
    pub sleeve_id: String,
    pub strategy_name: String,
    pub station_id: Option<String>,
    #[serde(default)]
    pub tickers: Vec<String>,
    pub market_type: Option<MarketType>,
    pub event_ticker: Option<String>,
    pub event_date: Option<NaiveDate>,
}

pub trait TimerHandle {
    fn cancelled(&self) -> bool;
    fn cancel(&mut self);
}

pub trait WorkHandle {
    type Error;

    fn cancelled(&self) -> bool;
    fn done(&self) -> bool;
    fn exception(&self) -> Option<&Self::Error>;
    fn cancel(&mut self);
}

pub trait EngineClock {
    fn now(&self) -> DateTime<Utc>;
    fn sleep(&self, seconds: f64) -> impl Future<Output = ()> + Send;
    fn sleep_until(&self, when: DateTime<Utc>) -> impl Future<Output = ()> + Send;
}

pub trait StrategyRuntime {
    type Clock: EngineClock;
    type Timer: TimerHandle;
    type Work: WorkHandle;

    fn mode(&self) -> RuntimeMode;
    fn run_id(&self) -> &str;
    fn scope(&self) -> &StrategyScope;
    fn clock(&self) -> &Self::Clock;
    fn runtime_identity(&self) -> &JsonValue;
    fn wake_at(&mut self, when: DateTime<Utc>, name: Option<&str>) -> Self::Timer;
    fn start_work<F>(&mut self, work: F, name: Option<&str>) -> Self::Work
    where
        F: Future<Output = ()> + Send + 'static;
}
